jxl/src/frame/modular/tree.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+use super::Predictor;
+use crate::{
+    bit_reader::BitReader,
+    entropy_coding::decode::Histograms,
+    error::{Error, Result},
+    util::tracing_wrappers::*,
+};
+
+#[allow(dead_code)]
+#[derive(Debug)]
+enum TreeNode {
+    Split {
+        property: u8,
+        val: i32,
+        left: u32,
+        right: u32,
+    },
+    Leaf {
+        predictor: Predictor,
+        offset: i32,
+        multiplier: u32,
+        id: u32,
+    },
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct Tree {
+    nodes: Vec<TreeNode>,
+    histograms: Histograms,
+}
+
+const SPLIT_VAL_CONTEXT: usize = 0;
+const PROPERTY_CONTEXT: usize = 1;
+const PREDICTOR_CONTEXT: usize = 2;
+const OFFSET_CONTEXT: usize = 3;
+const MULTIPLIER_LOG_CONTEXT: usize = 4;
+const MULTIPLIER_BITS_CONTEXT: usize = 5;
+const NUM_TREE_CONTEXTS: usize = 6;
+
+impl Tree {
+    #[instrument(level = "debug", skip(br), ret, err)]
+    pub fn read(br: &mut BitReader, size_limit: usize) -> Result<Tree> {
+        assert!(size_limit <= u32::MAX as usize);
+        trace!(pos = br.total_bits_read());
+        let tree_histograms = Histograms::decode(NUM_TREE_CONTEXTS, br, true)?;
+        let mut tree_reader = tree_histograms.make_reader(br)?;
+        // TODO(veluca): consider early-exiting for trees known to be infinite.
+        let mut tree: Vec<TreeNode> = vec![];
+        let mut to_decode = 1;
+        let mut leaf_id = 0;
+        let mut max_property = 0;
+        while to_decode > 0 {
+            if tree.len() > size_limit {
+                return Err(Error::TreeTooLarge(tree.len(), size_limit));
+            }
+            if tree.len() >= tree.capacity() {
+                tree.try_reserve(tree.len() * 2 + 1)?;
+            }
+            to_decode -= 1;
+            let property = tree_reader.read(br, PROPERTY_CONTEXT)?;
+            trace!(property);
+            if let Some(property) = property.checked_sub(1) {
+                // inner node.
+                if property > 255 {
+                    return Err(Error::InvalidProperty(property));
```

Comment:
Strictly speaking the spec does allow properties > 255, so this check is too strict. It would only make sense for images with a _lot_ of channels, and it wouldn't be something libjxl would ever produce, but in principle it is even legal to have a tree using property number 10000000 which is always always referencing some previous channel that doesn't exist so the value of the property is always zero.

It would probably make sense to add some sanity limits to the maximum property index that is allowed; e.g. limit it to 63 for Level 5 (which would be enough for 12 previous channels, since there are 16 fixed properties 0..15 and then 4 per prev-channel) and 255 for Level 10 (which is enough for 60 previous channels).

Reply:
This is the same check that libjxl has IIRC. I'd add it to the spec as a level 10 limit as you suggest.